apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/HiPerfPublisherTest.cs

```diff
@@ -0,0 +1,225 @@
+            private readonly IScheduler _schduler;
+            private readonly TaskHelper _taskHelper;
+            private readonly IEnumerator<TimeSpan> _delays;
+            private readonly List<string> _handledIds;
+
+            public IReadOnlyList<string> HandledMessages
+            {
+                get
+                {
+                    lock (_lock)
+                    {
+                        return _handledIds.ToArray();
+                    }
+                }
+            }
+
+            public override async Task<PublishResponse> PublishAsync(PublishRequest request, CallSettings callSettings = null)
+            {
+                _delays.MoveNext();
+                await _taskHelper.ConfigureAwait(_schduler.Delay(_delays.Current, callSettings?.CancellationToken ?? CancellationToken.None));
+                var msgIds = request.Messages.Select(x => x.Data.ToStringUtf8());
+                lock (_lock)
+                {
+                    _handledIds.AddRange(msgIds);
+                }
+                return new PublishResponse { MessageIds = { msgIds } };
+            }
+        }
+
+        private HiPerfPublisher.Settings MakeSettings(IList<PublisherClient> clients, IScheduler scheduler, int batchElementCountThreshold = 1, int batchRequestByteThreshold = 1)
+        {
+            return new HiPerfPublisher.Settings(clients)
+            {
+                Scheduler = scheduler,
+                BatchingSettings = new BatchingSettings
+                {
+                    DelayThreshold = TimeSpan.FromMilliseconds(10000),
+                    ElementCountThreshold = batchElementCountThreshold,
+                    RequestByteThreshold = batchRequestByteThreshold
+                }
+            };
+
+        }
+
+        [Theory, CombinatorialData]
+        public void OneMessage(
+            [CombinatorialValues(false, true)] bool hardStop)
+        {
+            var topicName = new TopicName("FakeProject", "FakeTopic");
+            var scheduler = new TestScheduler();
+            TaskHelper taskHelper = scheduler.TaskHelper;
+            var client = new FakePublisher(scheduler, taskHelper);
+            var settings = MakeSettings(new[] { client }, scheduler);
+            int shutdownCount = 0;
+            var pub = new HiPerfPublisherImpl(topicName, settings, () =>
+            {
+                shutdownCount += 1;
+                return Task.FromResult(0);
+            }, taskHelper);
+            scheduler.Run(async () =>
+            {
+                var id = await taskHelper.ConfigureAwait(pub.PublishAsync("1"));
+                bool isCancelled = await taskHelper.ConfigureAwaitHideCancellation(pub.ShutdownAsync(new CancellationToken(hardStop)));
+                Assert.Equal(hardStop, isCancelled);
+                Assert.Equal("1", id);
+                Assert.Equal(1, shutdownCount);
+            });
+        }
+
+        [Theory, PairwiseData]
```

Comment:
It ensures every possible pair of values are tested. Perhaps the [nunit documentation](https://github.com/nunit/docs/wiki/Pairwise-Attribute) is better.
It hugely reduces the combinatorial explosion of test-cases compared to using `CombinatorialData`.